components/org.wso2.carbon.configuration/src/main/java/org/wso2/carbon/configuration/ConfigProviderComponent.java

```diff
@@ -0,0 +1,100 @@
+import org.wso2.carbon.configuration.provider.ConfigProvider;
+import org.wso2.carbon.configuration.provider.ConfigProviderDataHolder;
+import org.wso2.carbon.configuration.provider.ConfigProviderImpl;
+import org.wso2.carbon.configuration.reader.ConfigFileReader;
+import org.wso2.carbon.configuration.reader.YAMLBasedConfigFileReader;
+import org.wso2.carbon.securevault.SecureVault;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * This service component is responsible for registering ConfigProvider OSGi service.
+ *
+ * @since 5.2.0
+ */
+@Component(
+        name = "org.wso2.carbon.configuration.ConfigProviderComponent",
+        immediate = true
+)
+public class ConfigProviderComponent {
+    private static final Logger logger = LoggerFactory.getLogger(ConfigProviderComponent.class);
+
+    @Activate
+    protected void activate(BundleContext bundleContext) {
+        ConfigProviderDataHolder.getInstance().setBundleContext(bundleContext);
+        initializeConfigProvider();
+        logger.debug("Activating Carbon Configuration Component");
+    }
+
+    @Deactivate
+    protected void deactivate(BundleContext bundleContext) {
+        ConfigProviderDataHolder.getInstance().setBundleContext(null);
+        logger.debug("Stopping ConfigProviderComponent");
+    }
+
+    @Reference(
+            name = "org.wso2.carbon.securevault",
+            service = SecureVault.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unRegisterSecureVault"
+    )
+    protected void registerSecureVault(SecureVault secureVault) {
+        ConfigProviderDataHolder.getInstance().setSecureVault(secureVault);
+        initializeConfigProvider();
+    }
+
+    protected void unRegisterSecureVault(SecureVault secureVault) {
+        ConfigProviderDataHolder.getInstance().setSecureVault(null);
+    }
+
+    /**
+     * Initialise carbon config provider.
+     */
+    private void initializeConfigProvider() {
+        if (!ConfigProviderDataHolder.getInstance().getSecureVault().isPresent() ||
+                !ConfigProviderDataHolder.getInstance().getBundleContext().isPresent()) {
+            logger.debug("Waiting for config provider dependencies");
+            return;
+        }
+        // TODO: Get carbon config location from carbon utils
+        // TODO: Support getting configProviderImpl in non-OSGi mode
+        Path deploymentConfigPath =
+                Paths.get(System.getProperty(Constants.CARBON_HOME), "conf", Constants.DEPLOYMENT_CONFIG_YAML);
+        ConfigFileReader configFileReader = new YAMLBasedConfigFileReader(deploymentConfigPath);
+        ConfigProvider configProvider = new ConfigProviderImpl(configFileReader);
+        ConfigProviderDataHolder.getInstance().getBundleContext()
+                .ifPresent(bundleContext -> bundleContext
+                        .registerService(ConfigProvider.class, configProvider, null));
```

Comment:
We set the bundle context in the `@Activate` method in the component. Therefore when this method is called bundle context may not be present